File: slim/sse.py
"""Incremental server-sent-events parsing. Stdlib only.

A socket chunk boundary can fall anywhere - inside a line, inside an event, in the
middle of the JSON - so the parser keeps a buffer instead of assuming one chunk is
one line or one event. Anything that does not start with "data:" (comments, the
event:/id:/retry: fields) is ignored, per the SSE spec.

`assemble()` exists because a server may answer with an event stream even when
stream was not requested; a plain "first content" extraction would then return only
the first delta.
"""
from __future__ import print_function

import json

DONE_MARKER = "[DONE]"


class SseParser(object):
    """Feed it raw bytes as text; pull complete data payloads out."""

    def __init__(self):
        self._buf = ""
        self._pending = []
        self._ready = []
        self._done = False

    def feed(self, chunk):
        if not chunk:
            return
        self._buf += chunk
        while True:
            idx = self._buf.find("\n")
            if idx < 0:
                break
            line = self._buf[:idx]
            self._buf = self._buf[idx + 1:]
            self._handle_line(line[:-1] if line.endswith("\r") else line)

    def flush(self):
        """End of stream: a server may omit the trailing newline / blank line.

        The leftover buffer is processed as the last line first, otherwise it would
        sit in the buffer and corrupt whatever is parsed next.
        """
        if self._buf:
            line, self._buf = self._buf, ""
            self._handle_line(line[:-1] if line.endswith("\r") else line)
        if self._pending:
            self._ready.append("\n".join(self._pending))
            self._pending = []

    def next_payload(self):
        """One complete data payload, or None when nothing is ready."""
        if not self._ready:
            return None
        return self._ready.pop(0)

    def done(self):
        """True once the server sent "[DONE]"."""
        return self._done

    def _handle_line(self, line):
        if line == "":
            if self._pending:
                self._ready.append("\n".join(self._pending))
                self._pending = []
            return
        if not line.startswith("data:"):
            return
        value = line[5:]
        if value.startswith(" "):
            value = value[1:]
        if value == DONE_MARKER:
            self._done = True
            return
        self._pending.append(value)


def delta_content(payload):
    """The assistant text in one payload, or None.

    choices[0].delta.content for a streaming chunk; choices[0].message.content if a
    server answered in one shot inside the stream.
    """
    if not payload:
        return None
    try:
        obj = json.loads(payload)
    except ValueError:
        return None
    choices = obj.get("choices") or []
    if not choices:
        return None
    first = choices[0] or {}
    for key in ("delta", "message"):
        part = first.get(key) or {}
        content = part.get("content")
        if content:
            return content
    return None


def assemble(raw_body):
    """Assemble a whole event-stream body (already in memory) into the reply text."""
    if not raw_body:
        return ""
    parser = SseParser()
    parser.feed(raw_body)
    parser.flush()
    parts = []
    while True:
        payload = parser.next_payload()
        if payload is None:
            break
        text = delta_content(payload)
        if text:
            parts.append(text)
    return "".join(parts)

File: slim/cpp/sse.cpp
#include "sse.hpp"

#include "util.hpp"

SseParser::SseParser() : done_(false) {}

void SseParser::reset() {
  buf_.clear();
  pending_.clear();
  ready_.clear();
  done_ = false;
}

void SseParser::feed(const std::string& chunk) {
  buf_ += chunk;
  std::string::size_type pos;
  while ((pos = buf_.find('\n')) != std::string::npos) {
    std::string line = buf_.substr(0, pos);
    buf_.erase(0, pos + 1);
    if (!line.empty() && line[line.size() - 1] == '\r')
      line.resize(line.size() - 1);  // tolerate CRLF
    handle_line(line);
  }
}

void SseParser::flush() {
  // A server may end the stream without a trailing newline or blank line. Process
  // whatever is still buffered as the final line first, otherwise that leftover
  // would sit in the buffer and corrupt whatever is parsed next.
  if (!buf_.empty()) {
    std::string line = buf_;
    buf_.clear();
    if (!line.empty() && line[line.size() - 1] == '\r')
      line.resize(line.size() - 1);
    handle_line(line);
  }
  if (!pending_.empty()) {
    ready_.push_back(pending_);
    pending_.clear();
  }
}

void SseParser::handle_line(const std::string& line) {
  if (line.empty()) {
    // Blank line terminates an event: whatever data lines were collected become
    // one payload (multi-line data joins with '\n', per the SSE spec).
    if (!pending_.empty()) {
      ready_.push_back(pending_);
      pending_.clear();
    }
    return;
  }
  if (line.compare(0, 5, "data:") != 0)
    return;  // comment (":"), event:, id:, retry:, or a line we do not speak
  std::string value = line.substr(5);
  if (!value.empty() && value[0] == ' ')
    value.erase(0, 1);
  if (value == "[DONE]") {
    done_ = true;
    return;
  }
  if (!pending_.empty())
    pending_ += "\n";
  pending_ += value;
}

bool SseParser::next(std::string* payload) {
  if (ready_.empty())
    return false;
  *payload = ready_.front();
  ready_.erase(ready_.begin());
  return true;
}

bool SseParser::done() const { return done_; }

bool sse_delta_content(const std::string& payload, std::string& out) {
  if (payload.empty())
    return false;
  if (json_extract_string(payload, "content", out))
    return !out.empty();
  return false;
}

bool sse_assemble(const std::string& raw_body, std::string& out) {
  out.clear();
  if (raw_body.empty())
    return false;
  SseParser parser;
  parser.feed(raw_body);
  parser.flush();
  std::string payload, delta;
  while (parser.next(&payload)) {
    if (sse_delta_content(payload, delta) && !delta.empty())
      out += delta;
  }
  return !out.empty();
}
